eyelid-match-ops/src/primitives/poly.rs

```diff
@@ -65,74 +86,247 @@ pub fn cyclotomic_mul(a: &Poly, b: &Poly) -> Poly {
 }
 
 /// Returns `a * b` followed by reduction mod `XˆN + 1` using recursive Karatsuba method.
-/// The returned polynomial has maximum degree [`MAX_POLY_DEGREE`].
-pub fn karatsuba_mul(a: &Poly, b: &Poly) -> Poly {
+/// The returned polynomial has a degree less than [`MAX_POLY_DEGREE`].
+///
+/// # Performance
+///
+/// This implementation should be compiled in release mode without debug checks.
+/// Some debug checks can slow it down by up to 25x:
+/// ```text
+/// debug = 2
+/// debug-assertions = true
+/// overflow-checks = true
+/// ```
+pub fn rec_karatsuba_mul(a: &Poly, b: &Poly) -> Poly {
+    rec_karatsuba_mul_inner(a, b, MAX_POLY_DEGREE)
+}
+
+/// Returns `a * b` followed by reduction mod `XˆN + 1` using recursive Karatsuba method.
+/// The returned polynomial has a degree less than or equal to `chunk`.
+///
+/// At each recusrsion level, polynomials start with maximum degree `chunk`, and are split to maximum degree `chunk/2`.
+fn rec_karatsuba_mul_inner(a: &Poly, b: &Poly, chunk: usize) -> Poly {
+    debug_assert!(a.degree() <= chunk);
+    debug_assert!(b.degree() <= chunk);
+
+    // invariant: the number of coefficients is a power of 2, before and after this function runs
+    debug_assert_eq!(chunk.count_ones(), 1);
+    const_assert_eq!(REC_KARATSUBA_MIN_DEGREE.count_ones(), 1);
+
     let mut res;
-    let n = a.degree() + 1;
 
     // if a or b has degree less than min, condition is true
-    let cond_a = a.degree() + 1 < MIN_KARATSUBA_REC_DEGREE;
-    let cond_b = b.degree() + 1 < MIN_KARATSUBA_REC_DEGREE;
+    let cond_a = a.degree() <= REC_KARATSUBA_MIN_DEGREE;
+    let cond_b = b.degree() <= REC_KARATSUBA_MIN_DEGREE;
     let rec_cond = cond_a || cond_b;
     if rec_cond {
         // If degree is less than the recursion minimum, just use the naive version
-        res = cyclotomic_mul(a, b);
+        res = a.naive_mul(b);
     } else {
+        // TODO: split this large code block into smaller functions, and benchmark the overall performance.
+        // (Smaller functions can be inlined, and the compiler can optimize better.)
+
         // Otherwise recursively call for al.bl and ar.br
-        let (al, ar) = poly_split(a);
-        let (bl, br) = poly_split(b);
-        let albl = karatsuba_mul(&al, &bl);
-        let arbr = karatsuba_mul(&ar, &br);
+        let (mut al, ar) = poly_split_half(a, chunk);
+        let (mut bl, br) = poly_split_half(b, chunk);
```

Comment:
I notice that this currently allocates new coefficient vectors for each of the polynomial parts -- in an ideal world, `al`, `ar`, `bl`, and `br` would be immutable slices and new memory would only be allocated for the `al + ar` and `bl + br` terms, computed below as `alpar` and `blpbr`, using only half of the memory.

This might be nontrivial to implement since Ark polynomials own their coefficients, but it seems possible that passing to coefficient vectors and slices might be worthwhile for a self-contained functionality like this.

Probably goes in the "performance optimizations" stack, however.

Reply:
Thanks, I opened ticket #51 for this.

Until we do #39, we won’t know if the memory usage or vector access is a significant part of the performance of the code. Rust vectors are highly optimised already, so it’s possible but unlikely.